### ISSUE TITLE
Think about hierarchies for test items

### ISSUE THREAD

**Issue by davidanthoff** [julia-testitems/TestItemRunner.jl#5]
At the moment the hierarchy for test items is essentially the file system, and then a flat list of `@testitem` per file.

Is that good enough? One really nice thing about this is that it is simple, plus we reuse an existing well established hierarchical structure, the file system, to order things. So right now I'm thinking that we probably don't want anything beyond that, but who knows...

One _could_ also consider something like `@testitem "Level 1/Level 2/My test" begin end`, and then the hierarchy would be constructed out of these paths of names...

**Comment by davidanthoff** [comment]
Another idea would be to add something like a `@testfolder` macro. Things might look like this:

```julia
@testfolder "Foo" begin
    @testfolder "Bar" begin
        @testitem "Something begin
        end
    end
end
```
The rule would presumably be that a `@testfolder` can _only_ contain other `@testfolder`s and `@testitem`s.

Could also be named `@testcollection`...

**Comment by jondea** [comment]
First let me say thank you for developing this package. It's one of those packages where you immediately think: I didn't know I needed this but I really do!

Reusing the file system does have a certain elegance and robustness.

You may have already considered this, but as an alternative could the existing `@testset` be repurposed for this? It is very natural to refer to a set of items, and the hierarchy in this system would reflect the default test runner.

**Comment by davidanthoff** [comment]
New idea: we conceptually change things so that each test item has a name/id and a label. The name/id can be hierarchical. If no label is specified, the name automatically becomes the label. The name/id can be specified either as a `String` or as a normal Julia identifier.

Implications would be this:

```julia
@testitem "Foo" begin
end
```

Here `"Foo"` is the id/name, and because no label is specified it also is the label.

```julia
@testitem Foo label="Something else" begin
end
```

Name is `Foo`, but label is "Something else"

```julia
@testitem Bar.Foo begin
end
```

The name is `Foo`, and the hierarchy will now show the file name, then a folder `Bar`, and `Foo` will appear inside the folder `Bar.

```julia
@testitem Bar.Foo label="Something else" begin
end
```

Would be the obvious thing.

And we would of course allow arbitrarily deep nesting:
```julia
@testitem Foobar.Bar.Foo begin
end
```

**Comment by lmiq** [comment]
Currently I have not felt the need of another level of nesting besides that of the files, and actually having the files as the top labels for the test makes it very easy to visualize where the tests are.

However, I do know that other packages have large files, and the list of tests can become complicated, thus I like the idea of allowing nesting by labeling. 

My doubt would be: If the `@testitem` is **not** nested, will it be at the top level, or as subitem of the "file"? Having to necessarily nest all testitems with a redundant filename information maybe a downgrade of usability. 

ps: Another syntax alternative could be:

```julia
@testgroup "Tests1"      # (actually `@testset` sounds nicer, but may cause conflicts).

@testitem "Foo" <: Tests1 begin 
end
```

That would not allow arbitrary nesting, but neither does our type system :-D.

**Comment by davidanthoff** [comment]
Ah, I should clarify: my idea is that the file/folder type nesting we have already stays in place, this would just be an optional way to add more nesting below the file system nesting we already have. I completely agree, I really like having the file system structure in place.

So this would also mean that if one has two files, `fileA.jl` and `fileB.jl`, and they have the following respective `@testitem`s in them:

In `fileA.jl`
```julia
@testitem Foo.Testitem1 begin
end
```
In fileB.jl`
```julia
@testitem Foo.Testitem2 begin
end
```
Then the hierarchy would look like this:
```
fileA.jl
  Foo
    Testitem1
fileB.jl
  Foo
    Testitem2
```
So the two `Foo` "folders" wouldn't be the same because they have different parents.

**Comment by cmichelenstrofer** [comment]
Maybe unrelated, but I think it would be great if the `Pkg.test` also showed even the current hierarchy. It currently lumps all `@testitem`s as a single testset named `Package`.

**Comment by NHDaly** [comment]
+1 for doing the nesting by directory/file structure. this maps to what they do in golang, and that seems to work well enough for them! :)

**Comment by NHDaly** [comment]
I'm ambivalent on the sub-file nesting, but it does seem like it could be nice, yeah! Either by having yet another grouping macro, or using nested names as you suggested.

**Comment by quinnj** [comment]
> my idea is that the file/folder type nesting we have already stays in place, this would just be an optional way to add more nesting below the file system nesting we already have. I completely agree, I really like having the file system structure in place.

Note that we don't seem to be doing actual nested folder -> nested tests yet; for example, if I have 2 subdirectories in my test directory, each w/ their own `@testitem`s, it shows up like:

<img width="543" alt="image" src="https://user-images.githubusercontent.com/2896623/212149728-dadb48c9-2053-4f23-a372-361a299b789d.png">

i.e. each individual file w/ `@testitem`s is listed as a top-level "set" of tests

Whereas in go (which @NHDaly was referring to), has:

<img width="535" alt="image" src="https://user-images.githubusercontent.com/2896623/212150018-679437c1-0291-46d8-9f18-69b5ff60319d.png">

so here we have the `std` directory, in which we have the `bytes` directory, and finally the `boundary_test.go` file has a nested list of individual tests.

As far as I can tell, we need to restructure how we're providing testitems to the vscode request by utilizing the [`children`](https://github.com/microsoft/vscode/blob/638db267973e2fb8d4c9e02372340f37a1a11f25/src/vscode-dts/vscode.d.ts#L16268) field to have the appropriate nesting show up.

If I can figure out how to get a dev env setup to hack on the julia extension, I'll see if I can take a crack at making this change.

**Comment by davidanthoff** [comment]
> Note that we don't seem to be doing actual nested folder -> nested tests yet

Yep, I was more describing an aspiration ;) Not yet implemented, and definitely the `go` style is better as one can then run all tests in a given folder.

I think this folder hierarchy thing could be implemented purely in the VS Code extension, essentially https://github.com/julia-vscode/julia-vscode/blob/d0d8c6b43fc7cf3f96cd41493e7a943829c421e1/src/testing/testFeature.ts#L295-L328 would just need to parse `params.uri` into the individual folder parts and then construct the correct hierarchy of `TestItem`s. The implementation might be a bit annoying because the `publishTestitemsHandler` notification is per file, so one then needs to go into the tree of `TestItem`s and only modify parts of it. I'm also not entirely sure whether we handle deletion of files properly (even in the current implementation).

And we might then also need to check https://github.com/julia-vscode/julia-vscode/blob/d0d8c6b43fc7cf3f96cd41493e7a943829c421e1/src/testing/testFeature.ts#L414 to make sure it can deal with `TestItem`s that are folders, but I think that should be fairly straightforward.

**Comment by MilesCranmer** [comment]
Any progress on this? I wanted to do

```julia
@testitem "Set of tests" begin
    f(x) = x
    @testitem "Single part" begin
        import ..f
        f(1)
    end
end
```

but this didn't seem to work. I think according to https://en.wikipedia.org/wiki/Principle_of_least_astonishment it would be nice to have this since it's possible with `testset`, and intuitively I think of the `@testitem` as defining little modules.

**Comment by davidanthoff** [comment]
So, I don't think we would ever allow nesting of `@testitem`, they are really the atomic leaf unit that one can execute. If we were to enable some in-file nesting it would presumably use one of the solutions that we discussed above.

But for now I think the low hanging fruit is to just properly implement a tree view for the folder/file hierarchy in the extension, that seems the first step.

**Comment by moble** [comment]
One feature I use a lot — and is actually stopping me from converting one of my packages to use TestItems — is the ability to generate a series of `@testset`s with a for loop.  My main use case is to run the same tests with a variety of types.  For example, here's a particularly simple testset that I just got annoyed at for not showing up nicely in your wonderful interface:
```julia
@testset verbose=true "Align QuatVec{$T}" for T in [Float16, Float32, Float64]
```
I'll usually have much more complicated types to loop over, and seeing the results as separate tests lets me focus in on the source of failures more quickly.

That's the only pain point I encounter when using this otherwise amazing package, so thank you very much for all your work!

**Comment by nickrobinson251** [comment]
FWIW at my workplace (RAI) we migrated from `@testset` -> `@testitem`, and have a test suite of ~5000 testitems, and have survived fine without the `for`-loop syntax, and are actually please it's not available (at least in a way that could add latency).

For the most part we just converted these cases by moving the `for`-loop inside, like
```julia
@testitem "Align QuatVec" begin
  @testset verbose=true "Align QuatVec{$T}" for T in [Float16, Float32, Float64]
    ...
  end
end
```
or in a few cases (where the tests were slow but not so many of them) we  manually wrote them out like, possibly using `@testsetup` to reduce duplicated boilerplate, like
```
@testsetup QuatVecSetup
 ...
end
@testitem "Align QuatVec{Float16}" setup=[QuatVecSetup] begin
  ...
end 
@testitem "Align QuatVec{Float32}" setup=[QuatVecSetup] begin
  ...
end
```
and in one case we did right a script to generate 100s of `@testitems` in a `for`-loop and committed the result (and script).

While the migration was a little bit annoying, because we had to decide what to do for each `@testset for` case if "moving it inside" made the tests too slow, we actually like the result. Otherwise, with `@testitem for ...` i think you'd either have to generate these testitems up front before the tests can run, meaning their can be arbitrary latency before tests start (slowing down the whole test suite), or else i can't see how you'd be able to run singular testitems from the loop by name (e.g. just `Align QuatVec{Float32}`) at which point they may as well be inside a single testitem.

**Comment by davidanthoff** [comment]
I'm closing this, I think for now we have enough hiearchies :)
